pkg/backup/backup.go

```diff
@@ -174,94 +175,186 @@ func (b *Backup) Run() {
+		return lastSnapRev, fmt.Errorf("failed to write snapshot to backend: (%v)", err)
 	}
 
-	log.Printf("saving backup for cluster (%s)", b.clusterName)
-	if err := b.writeSnap(member, rev); err != nil {
-		err = fmt.Errorf("write snapshot failed: %v", err)
-		return lastSnapRev, err
+	bs := backupapi.BackupStatus{
+		CreationTime:     time.Now().Format(time.RFC3339),
+		Size:             util.ToMB(n),
+		Version:          etcdVersion,
+		Revision:         rev,
+		TimeTookInSecond: int(time.Since(start).Seconds() + 1),
 	}
+	b.recentBackupsStatus = append(b.recentBackupsStatus, bs)
+	if len(b.recentBackupsStatus) > maxRecentBackupStatusCount {
+		b.recentBackupsStatus = b.recentBackupsStatus[1:]
+	}
+
 	return rev, nil
 }
 
-func (b *Backup) writeSnap(m *etcdutil.Member, rev int64) error {
-	start := time.Now()
+// WriteSnap writes the snapshot to the backend.
+func WriteSnap(version string, rev int64, be backend.Backend, snap *Snap) (int64, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), constants.DefaultSnapshotTimeout)
+	defer cancel()
+	rc, err := snap.Snapshot(ctx)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get snapshot (%v)", err)
+	}
+
+	n, err := be.Save(version, rev, rc)
+	rc.Close()
+	if err != nil {
+		return 0, err
+	}
+	return n, nil
+}
+
+// GetSnap gets the snapshot from the cluster member with the maximum revision.
+func GetSnap(kclient kubernetes.Interface, tls *tls.Config, namespace, clusterName string) (*Snap, int64, error) {
+	pods, err := getRunningPods(kclient, namespace, clusterName)
+	if err != nil {
+		return nil, 0, fmt.Errorf("failed to retrieve running etcd pods: (%v)", err)
+	}
+	if len(pods) == 0 {
+		return nil, 0, fmt.Errorf("no running etcd pods found")
+	}
+
+	m, rev := getMemberWithMaxRev(pods, tls)
+	if m == nil {
+		return nil, 0, fmt.Errorf("no reachable member")
+	}
+
+	return NewSnap(m.ClientURL(), tls), rev, nil
+}
 
+func NewSnap(url string, tls *tls.Config) *Snap {
 	cfg := clientv3.Config{
-		Endpoints:   []string{m.ClientURL()},
+		Endpoints:   []string{url},
 		DialTimeout: constants.DefaultDialTimeout,
-		TLS:         b.etcdTLSConfig,
+		TLS:         tls,
 	}
-	etcdcli, err := clientv3.New(cfg)
+	return &Snap{cfg: cfg}
+}
+
+type Snap struct {
+	cfg clientv3.Config
+	// mu protects lazy creation of cli.
+	mu sync.Mutex
+	// cli is lazily created when either Snapshot or Version functions
+	// are being called.
+	cli *clientv3.Client
+}
+
+// Snapshot returns a snapshot from the etcd member.
+func (s *Snap) Snapshot(ctx context.Context) (io.ReadCloser, error) {
+	cli, err := s.lazyClient()
+	if err != nil {
+		return nil, err
+	}
+	rc, err := cli.Maintenance.Snapshot(ctx)
 	if err != nil {
-		return fmt.Errorf("failed to create etcd client (%v)", err)
+		return nil, fmt.Errorf("failed to get member snapshot (%v)", err)
 	}
-	defer etcdcli.Close()
+	return rc, nil
+}
 
+// Version returns the etcd version of the etcd member.
+func (s *Snap) Version() (string, error) {
```

Comment:
this method does not make sense. why snapshot.version return member's version not snapshot's version? it seems too confusing.

Reply:
Maybe change method name to `EtcdVersion()` or remove`Version()` in side snap struct?

Reply:
probably https://github.com/coreos/etcd-operator/pull/1489#discussion_r143321802 needs to be answered first.